File: Mozu.SiteBuilder.UX.Admin/Api/Models/Attributes/Attribute.cs
using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Mozu.SiteBuilder.UX.Admin.Api.Models.Attributes
{
    [DataContract]
    public class Attribute
    {
        [DataMember(Name = "id")]
        public string Id { get; set; }

        [DataMember(Name = "name")]
        public string Name { get; set; }

         [JsonConverter(typeof(StringEnumConverter))]
        [DataMember(Name = "inputType")]
        public AttributeInputType InputType { get; set; }

         [JsonConverter(typeof(StringEnumConverter))]
        [DataMember(Name = "dataType")]
        public AttributeDataType DataType { get; set; }


        [DataMember(Name = "isOption")]
        public bool IsOption { get; set; }

        [DataMember(Name = "isExtra")]
        public bool? IsExtra { get; set; }

        [DataMember(Name = "isProperty")]
        public bool? IsProperty { get; set; }


        [DataMember(Name = "valueType")]
        [JsonConverter(typeof(StringEnumConverter))]
        public AttributeValueType ValueType { get; set; }


        [DataMember(Name = "attributeMetadata")]
        public List<AttributeMetadataItem> AttributeMetadata { get; set; }

        

 

        [DataMember(Name = "min")]
        public decimal? Min { get; set; }

        [DataMember(Name = "minDate")]
        public DateTime? MinDate { get; set; }

        [DataMember(Name = "max")]
        public decimal? Max { get; set; }

        [DataMember(Name = "maxDate")]
        public DateTime? MaxDate { get; set; }

        [DataMember(Name = "regex")]
        public string Regex { get; set; }

         [DataMember(Name = "adminName")]
        public string AdminName { get; set; }

        [DataMember(Name = "values")]
        public List<AttributeValue> Values { get; set; }


    }


    [DataContract]
    public class AttributeMetadataItem
    {
        [DataMember(EmitDefaultValue = false, Name="key")]
        public string Key { get; set; }

        [DataMember(EmitDefaultValue = false, Name = "value")]
        public string Value { get; set; }
    }
}

File: Mozu.SiteBuilder.UX.Admin/Scripts/app/model/Attribute.js
/**
 * @class Taco.model.Attribute
 */
Ext.define('Taco.model.Attribute', {
    requires: ['Taco.model.AttributeValue'],
    extend: 'Taco.core.data.Model',
    fields: [{
            name: 'id',
            type: 'auto'
        }, {
            /* numeric id of this attribute used by the service */
            name: 'attributeId',
            type: 'int'
        }, {
            name: 'name',
            type: 'string'
        }, {
            name: 'code',
            type: 'string',
            useNull: true
        }, {
            name: 'inputType',
            type: 'string'
        }, {
            name: 'dataType',
            type: 'string'
        }, {
            name: 'valueType',
            type: 'string'
        }, {
            name: 'isOption',
            type: 'boolean'
        }, {
            name: 'isExtra',
            type: 'boolean'
        }, {
            name: 'adminName',
            type: 'string'
        }, {
            name: 'isActive',
            type: 'boolean'
        }, {
            name: 'isRequired',
            type: 'boolean'
        },
        {
            name:'isVisible',
            type:'boolean',
            defaultValue:false
        },
        {
            name: 'displayGroup',
            type: 'string',
            defaultValue: 'Admin'
        }, {
            name: 'isProperty',
            type: 'boolean'
        }, {
            name: 'min',
            type: 'auto',
            convert: function (v, r) {
                return r.convert(v, r);
            }
        }, {
            name: 'max',
            type: 'auto',
            convert: function (v, r) {
                return r.convert(v, r);
            }
        }, {
            name: 'minDate',
            type: 'auto',
            convert: function (v, r) {
                return r.convert(v, r);
            }
        }, {
            name: 'maxDate',
            type: 'auto',
            convert: function (v, r) {
                return r.convert(v, r);
            }
        }, {
            name: 'rows',
            type: 'auto'
        }, {
            name: 'attributeMetadata',
            type: 'auto',
            defaultValue: []
        }, {
            name: 'regex',
            type: 'string'
        }, {
            name: 'values',
            type: 'auto',
            defaultValue: []
        }
    ],
    supportsAttributeType:function () {
        return true;
    },
    supportsDisplayGroup: function () {
        return false;
    },
    supportsValueType: function () {
        return false;
    },
    supportsIsRequired: function () {
        return false;
    },
    supportsIsVisible: function () {
        return false;
    },
    allowProductDataType: function () {
        return true;
    },

    convert: function (v, r) {
        if (r.data.dataType == 'DateTime') {
            return r.convertDate(v);
        }
        if (r.data.dataType == 'Bool') {
            return r.convertBool(v);
        }
        if (r.data.dataType == 'Number') {
            return r.convertNumber(v);
        }

        return v;
    },
    convertBool: function (v) {
        if ((v === undefined || v === null || v === '')) {
            return null;
        }
        return v === true || v === 'true' || v == 1;
    },
    convertDate: function (v) {

        if (!v) {
            return null;
        }
        if (Ext.isDate(v)) {
            return v;
        }
        return Ext.Date.parse(v, 'c');

    },
    convertNumber: function (v) {
        return v !== undefined && v !== null && v !== '' ?
            parseFloat(String(v).replace(Ext.data.Types.stripRe, ''), 10) : null;
    },
    getAttributeValues: function () {
        return this.getOrCreateHasManyStore({
            model: 'Taco.model.AttributeValue',
            associationKey: 'values',
            foreignKey: 'attributeId',
            foreignProperty: 'attribute'
        });


    },
    validations: [{
            type: 'length',
            name: 'name',
            min: 3,
            max: 100
        }, {
            type: 'presence',
            name: 'name'
        }, {
            type: 'length',
            name: 'adminName',
            min: 3,
            max: 100
        }, {
            type: 'presence',
            name: 'adminName'
        }
        // { type: 'format', name: 'productCode', matcher: /^[A-z0-9\-]*$/ }
    ],
    proxy: {
        type: 'ajaxproxy',
        // api: {
        //     create: '/admin/app/Test/testCreate',
        //     read: '/admin/Scripts/app/mocks/attributes.json',
        //     update: '/admin/app/Test/testUpdate',
        //     destroy: '/admin/app/Test/testDestroy'
        // },
        api: {
            create: '/admin/app/attribute/create',
            read: '/admin/app/attribute/read',
            update: '/admin/app/attribute/update',
            destroy: '/admin/app/attribute/destroy'
        },
        reader: {
            type: 'json',
            root: 'items',
            successProperty: 'success'
        },
        writer: {
            type: 'json',
            allowSingle: false
        }
    }
});